### PR TITLE
Provide support for yaml constructors !degrees and !radians

### PR DESCRIPTION
**Pull request by rhaschke**
Addresses #251. @gavanderhoorn, please review.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
This is :100: again @rhaschke.

My intention with #251 was not to come up with more work for you, but I'm of course very happy you decided to implement it.

I'll run some tests and let you know.

**Review by gavanderhoorn — APPROVED** [review]
This works for me :+1: 

thanks for adding this @rhaschke.

File: src/xacro/__init__.py [hunk 1]
```diff
@@ -101,9 +101,31 @@ def __getattr__(self, item):
...
+        return float(eval(expr))
+    except SyntaxError as e:
+        raise XacroException("invalid expression: %s" % value)
+
+
+def construct_angle_degrees(loader, node):
+    """utility function for converting degrees into radians from yaml"""
+    value = loader.construct_scalar(node)
+    try:
+        return float(value) * math.pi / 180.0
```
Comment (gavanderhoorn) [inline src/xacro/__init__.py]:
```suggestion
        return math.radians(float(value))
```

**Comment by gavanderhoorn** [comment]
Do I understand the code correctly when I say `!radians` appears to support "arbitrary expressions", while `!degrees` doesn't?

**Comment by rhaschke** [comment]
> Do I understand the code correctly when I say `!radians` appears to support "arbitrary expressions", while `!degrees` doesn't?

Yes, that's what I copied from [ros_param](https://github.com/ros/ros_comm/blob/12eaf946597739e833b2300ef28bb079f248a9e1/tools/rosparam/src/rosparam/__init__.py#L635-L646). Do you want to change this?

**Comment by gavanderhoorn** [comment]
> Yes, that's what I copied from [ros_param](https://github.com/ros/ros_comm/blob/12eaf946597739e833b2300ef28bb079f248a9e1/tools/rosparam/src/rosparam/__init__.py#L635-L646). Do you want to change this?

We should probably keep the same restrictions as in `rosparam`, so I would vote to keep it as-is.

Only thing to make sure perhaps is the `eval(..)` there: does it open up `xacro` to security issues? Should we limit / sanatise something perhaps?

**Comment by rhaschke** [comment]
Indeed, we should restrict `eval()` to our limited set of functions...

File: src/xacro/__init__.py [hunk 2]
```diff
@@ -101,9 +101,30 @@ def __getattr__(self, item):
 
     __getitem__ = __getattr__
 
+
+def construct_angle_radians(loader, node):
+    """utility function to construct radian values from yaml"""
+    value = loader.construct_scalar(node).strip()
+    try:
+        return float(eval(expr, global_symbols, dict(pi=math.pi)))
```
Comment (gavanderhoorn) [inline src/xacro/__init__.py]:
```suggestion
        return float(eval(value, global_symbols, dict(pi=math.pi)))
```
?

Comment (gavanderhoorn) [reply src/xacro/__init__.py]: Or `s/value/expr/g` of course.

**Comment by gavanderhoorn** [comment]
@rhaschke: just submitted ubi-agni/xacro#4, which adds some (primitive) testing to the safe(r) `eval(..)` functionality when used in the yaml ctors.

**Comment by gavanderhoorn** [comment]
Thanks @rhaschke :+1:

**Comment by gavanderhoorn** [comment]
@rhaschke: would you have time to release a new version of xacro which includes this change?

**Comment by rhaschke** [comment]
Done: https://github.com/ros/rosdistro/pull/26352

**Comment by gavanderhoorn** [comment]
@rhaschke: I'm not sure I remember your syncing/back-porting policy between branches here.

Would it be possible to get this in Noetic as well?

**Comment by rhaschke** [comment]
Sure. I will prepare a new release later this evening.

**Comment by rhaschke** [comment]
Done: https://github.com/ros/rosdistro/pull/26945

**Comment by gavanderhoorn** [comment]
Thanks @rhaschke :+1:

**Comment by Briancbn** [comment]
Will this be merged into foxy release? I have to merge this to get the melodic-devel-staging urdf models working in ROS2.

**Comment by rhaschke** [comment]
@Briancbn: Sure. Do you mind preparing a merge PR? This would help to speed up the process.

**Comment by Briancbn** [comment]
@rhaschke could you help review the PR? thanks.

**Comment by rhaschke** [comment]
#259 is on my todo list. But there is just too much other work...

**Pull request merged**